### PR TITLE
feat: new UDFs for set-like operations on Arrays

### PR DESCRIPTION
**Pull request by blueedgenick**
Four new UDFs, implementing set-like operations on Array fields,inspired by the similarly-named functions in Presto (and others):

- `array_distinct(array)`
- `array_except(array1, array2)`
- `array_intersect(array1, array2)`
- `array_union(array1, array2)`

This is a revised and updated version of these functions and their tests, which were originally proposed long ago  in PR #1611 and blocked at that time by limited scope for handling collections of generics in the UDF invocation framework. 

Review feedback from the original PR incorporated herein, along with QTTs and historical plans.

### CODE REVIEW

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayExceptTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.nullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+
+import java.util.Arrays;
+import java.util.List;
+import org.junit.Test;
+
+public class ArrayExceptTest {
```
Comment (vpapavas) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayExceptTest.java:26]: You are missing a test case where the arrays contain null values

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayUnionTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    final List<String> result = udf.union(input1, input2);
+    assertThat(result, contains((String) null, "foo", "bar"));
+  }
+
+  @Test
+  public void shouldReturnNullForArraysOfOnlyNulls() {
+    final List<String> input1 = Arrays.asList(null, null);
+    final List<String> input2 = Arrays.asList(null, null, null);
+    final List<String> result = udf.union(input1, input2);
+    assertThat(result, contains(nullValue()));
```
Comment (vpapavas) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayUnionTest.java:81]: This is a bit confusing behavior for me. Why does the test right above contain a null value in the resulting array but this test returns null?

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayExceptTest.java (hunk 1 above)
Comment (blueedgenick) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayExceptTest.java:26]: argh, good catch - thanks! will add

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayUnionTest.java (hunk 2 above)
Comment (blueedgenick) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayUnionTest.java:81]: there's a subtlety here - in this case the udf doesn't really return null, instead it returns _an array containing a single null_ - which i think is what the function should do in this case, on the grounds that the only distinct value found in either of the input arrays is a null.  This is different than the case that we pass in an actual null, as compared to array-of-nulls which is what we have here, and that other case is tested below.
Does that make sense, or am i misunderstanding the question ?

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 3]
```diff
@@ -156,6 +156,55 @@ Given an array, checks if a search value is contained in the array.
...
 Accepts any `ARRAY` type. The type of the second param must match the element type of the `ARRAY`.
 
+### ``ARRAY_DISTINCT``
+
+```sql
+ARRAY_DISTINCT([1, 2, 3])
+```
+
+Returns an array of all the distinct values, including NULL if present, from the input array.
+The output array elements will be in order of their first occurrence in the input.
```
Comment (JimGalasyn) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:166]:
```suggestion
The output array elements are in order of their first occurrence in the input.
```

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 4]
```diff
@@ -156,6 +156,55 @@ Given an array, checks if a search value is contained in the array.
...
+ARRAY_DISTINCT(ARRAY['apple', 'apple', NULL, 'cherry'])  => ['apple', NULL, 'cherry']
+```
+
+### ``ARRAY_EXCEPT``
+
+```sql
+ARRAY_EXCEPT(array1, array2)
+```
+
+Returns an array of all the distinct elements from an array except for those also present in a second array. The order of entries in the first array is preserved although any duplicates are removed.
```
Comment (JimGalasyn) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:182]:
```suggestion
Returns an array of all the distinct elements from an array, except for those also present in a second array. The order of entries in the first array is preserved, but duplicates are removed. 
```

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayUnionTest.java (hunk 2 above)
Comment (vpapavas) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayUnionTest.java:81]: Oh sorry, I just noticed the `contains`. I thought it was `equals` that's why I asked. Your answer and the code totally make sense. Sorry about that

Comment (blueedgenick) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayUnionTest.java:81]: yeah, this new matcher-naming throws me off regularly too - i think i liked "contains" much better as something like "isArrayContaining", especially when you're scanning through a whole set of tests  :)

**Pull request merged**